### PR TITLE
Fix estimate gas logic

### LINKED ISSUE #7: Fix estimate gas logic
### Expected behavior
Transaction should be posted successfully to the network without any issues

### Actual behavior
Currently, `estimateGas` method returns lower gas which is causing transaction to fail

### Steps to reproduce
- Start the application and post transaction

### PR DESCRIPTION
**Pull request by nagdahimanshu**
### What was the problem?

This PR resolves #7 

### How was it solved?

- [x] Update gas estimation logic to use gasLimit
  - [x] 35000 for initialised accounts
  - [x] 55000 for non-initialised accounts
- [x] Update Makefile to include command to generate Go bindings of smart contract
- [x] Update Readme 

### How was it tested?
Local

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -4,25 +4,36 @@ PKGS=$(shell go list ./... | grep -v "/vendor/")
...
 	@echo "$(BLUE)» Building frontend... $(COLOR_END)"
-	@go generate
+	go generate
 	@echo "$(GREEN)Frontend successfully built$(COLOR_END)"
 
 run: # Runs the application, use `make run FLAGS="--help"`
-	@./bin/${APP_NAME} ${FLAGS}
+	./bin/${APP_NAME} ${FLAGS}
+
+generate-abi:
```
Comment (sameersubudhi) [inline Makefile]: Let's also add a env variable validation here such that the user is informed that they need to pass the `ERC20_CONTRACT_FILE_PATH` variable.

File: README.md [hunk 2]
```diff
@@ -22,6 +22,8 @@ LSK faucet is a web application that can be configured and deployed to get ETH a
 
 * Go (1.22 or later)
 * Node.js
+* Solidity compiler (Solc)
```
Comment (sameersubudhi) [inline README.md]: Let's also specify that the compiler version needs to match the contract version 🙏 
```suggestion
* Solidity compiler (solc)
```

File: internal/server/server.go [hunk 3]
```diff
@@ -50,12 +52,18 @@ func (s *Server) handleClaim() http.HandlerFunc {
...
 		address, _ := readAddress(r)
 		ctx, cancel := context.WithTimeout(r.Context(), 5*time.Second)
 		defer cancel()
-		txHash, err := s.TransferERC20(ctx, address, chain.LSKToWei(int64(s.cfg.payout)))
+
+		currBalance, err := s.GetContractInstance().BalanceOf(&bind.CallOpts{}, common.HexToAddress(address))
+		if err != nil {
+			log.WithError(err).Error("Failed to fetch recipient balance")
+			return
+		}
```
Comment (sameersubudhi) [inline internal/server/server.go]: Instead of returning error here, let's log the error and assume that the account doesn't exist and proceed with the ERC20 transfer with the higher gas limit.

File: internal/chain/transaction_test.go [hunk 4]
```diff
@@ -78,17 +78,18 @@ func TestTxBuilder_TransferERC20(t *testing.T) {
...
+		fromAddress:      crypto.PubkeyToAddress(privateKey.PublicKey),
+		chainID:          big.NewInt(1337),
+		tokenAddress:     "0xbb5801a7D398351b8bE11C439e05C5B3259aeux0",
+		contractInstance: nil,
 	}
 	bgCtx := context.Background()
 	toAddress := common.HexToAddress("0xAb5801a7D398351b8bE11C439e05C5B3259aeC9B")
 	value := big.NewInt(1000)
-	txHash, err := txBuilder.TransferERC20(bgCtx, toAddress.Hex(), value)
+	txHash, err := txBuilder.TransferERC20(bgCtx, toAddress.Hex(), value, big.NewInt(30000))
```
Comment (sameersubudhi) [inline internal/chain/transaction_test.go]: Can we also add another test case for the zero balance account and verify the used gas limits for the transaction?
Please add the gas limit verification logic to this test case as well.

File: internal/chain/transaction.go [hunk 5]
```diff
@@ -131,12 +143,11 @@ func (b *TxBuild) TransferERC20(ctx context.Context, to string, value *big.Int)
...
-		To:   &tokenAddress,
-		Data: data,
-	})
-	if err != nil {
-		return emptyHash, err
+	var gasLimit uint64
+	if balance.BitLen() == 0 {
+		gasLimit = 55000
+	} else {
+		gasLimit = 35000
```
Comment (sameersubudhi) [inline internal/chain/transaction.go]: Can we define constants for these gas limits so that they can be reused within the unit tests as well?

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: README.md [hunk 6]
```diff
@@ -38,8 +40,14 @@ cd lsk-faucet
 ```bash
 make build-frontend
 ```
+3. Generate Go binding for ERC20 token smart contract
+```bash
+make generate-binding ERC20_CONTRACT_FILE_PATH=<erc20-contract-file-path>
+```
+
+**NOTE**: Please make sure to generate Go binding before building the Go project.
```
Comment (sameersubudhi) [inline README.md]:
```suggestion
**NOTE**: Please make sure to generate the Go binding corresponding to your contract before building the project.
```

File: internal/chain/transaction_test.go [hunk 7]
```diff
@@ -78,17 +79,20 @@ func TestTxBuilder_TransferERC20(t *testing.T) {
...
+		chainID:          big.NewInt(1337),
+		tokenAddress:     "0xbb5801a7D398351b8bE11C439e05C5B3259aeux0",
+		contractInstance: nil,
 	}
 	bgCtx := context.Background()
 	toAddress := common.HexToAddress("0xAb5801a7D398351b8bE11C439e05C5B3259aeC9B")
 	value := big.NewInt(1000)
-	txHash, err := txBuilder.TransferERC20(bgCtx, toAddress.Hex(), value)
+
+	// Transaction when recipient is initialized
```
Comment (sameersubudhi) [inline internal/chain/transaction_test.go]: Both the scenarios should be tested as independent testcases. 
```
func TestTxBuilder_TransferERC20(t *testing.T) {
	testcases := []struct {
		name         string
		accBalance   *big.Int
		wantGasLimit uint64
	}{
		{
			name:         "should use gasLimitInitializedAccount gas limit for transaction when recipient is initialized",
			accBalance:   big.NewInt(30000),
			wantGasLimit: gasLimitInitializedAccount,
		},
		{
			name:         "should use gasLimitNonInitializedAccount gas limit for transaction when recipient is not initialized",
			accBalance:   big.NewInt(0),
			wantGasLimit: gasLimitNonInitializedAccount,
		},
	}

	t.Parallel()
	for _, tc := range testcases {
		t.Run(tc.name, func(t *testing.T) {
			privateKey, _ := crypto.HexToECDSA("976f9f7772781ff6d1c93941129d417c49a209c674056a3cf5e27e225ee55fa8")
			fromAddress := crypto.PubkeyToAddress(privateKey.PublicKey)
			simBackend := simulated.NewBackend(
				types.GenesisAlloc{
					fromAddress: {Balance: big.NewInt(10000000000000000)},
				})

			defer simBackend.Close()
			var s *backends.SimulatedBackend
			patches := gomonkey.ApplyMethod(reflect.TypeOf(s), "SuggestGasPrice", func(_ *backends.SimulatedBackend, _ context.Context) (*big.Int, error) {
				return big.NewInt(875000000), nil
			})
			defer patches.Reset()

			txBuilder := &TxBuild{
				client:           simBackend.Client(),
				privateKey:       privateKey,
				signer:           types.NewEIP155Signer(big.NewInt(1337)),
				fromAddress:      crypto.PubkeyToAddress(privateKey.PublicKey),
				chainID:          big.NewInt(1337),
				tokenAddress:     "0xbb5801a7D398351b8bE11C439e05C5B3259aeux0",
				contractInstance: nil,
			}
			bgCtx := context.Background()
			toAddress := common.HexToAddress("0xAb5801a7D398351b8bE11C439e05C5B3259aeC9B")
			value := big.NewInt(1000)

			txHashInitAcc, err := txBuilder.TransferERC20(bgCtx, toAddress.Hex(), value, tc.accBalance)
			if err != nil {
				t.Errorf("could not add tx to pending block: %v", err)
			}
			simBackend.Commit()

			block, err := simBackend.Client().BlockByNumber(bgCtx, big.NewInt(1))
			if err != nil {
				t.Errorf("could not get block at height 1: %v", err)
			}
			if txHashInitAcc != block.Transactions()[0].Hash() {
				t.Errorf("did not commit sent transaction. expected hash %v got hash %v", block.Transactions()[0].Hash(), txHashInitAcc)
			}

			txInitAcc, _, err := simBackend.Client().TransactionByHash(bgCtx, txHashInitAcc)
			if err != nil {
				t.Errorf("could not get transaction by hash: %v", err)
			}

			gasLimit := txInitAcc.Gas()
			assert.Equal(t, gasLimit, tc.wantGasLimit)

			// TODO: Verify ERC20 token balance of recipient if possible in future with the simulator
		})
	}
}
```

File: internal/server/server.go [hunk 8]
```diff
@@ -50,12 +53,18 @@ func (s *Server) handleClaim() http.HandlerFunc {
...
-
 		// The error always be nil since it has already been handled in limiter
 		address, _ := readAddress(r)
 		ctx, cancel := context.WithTimeout(r.Context(), 5*time.Second)
 		defer cancel()
-		txHash, err := s.TransferERC20(ctx, address, chain.LSKToWei(int64(s.cfg.payout)))
+
+		currBalance, err := s.GetContractInstance().BalanceOf(&bind.CallOpts{}, common.HexToAddress(address))
+		if err != nil {
+			log.WithError(err).Error("Failed to fetch recipient balance")
```
Comment (sameersubudhi) [inline internal/server/server.go]:
```suggestion
			log.WithError(err).Error("failed to fetch recipient balance")
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]
Nice job so far, please check the following recommendations 🙏

**Review by sameersubudhi — APPROVED** [review]

**Review by hrmhatef — APPROVED** [review]
LGTM👍
